connectivity/check/action.go

```diff
@@ -510,197 +549,198 @@ func (a *Action) GetIngressRequirements(p FlowParameters) []filters.FlowSetRequi
-
-retryOnEmpty:
-	w := utils.NewWaitObserver(ctx, utils.WaitParameters{
-		Timeout:         defaults.FlowWaitTimeout,
-		RetryInterval:   defaults.FlowRetryInterval,
-		WarningInterval: defaults.FlowWaitTimeout / 2, // warn at least once during wait timeout
-		Log: func(err error, wait string) {
-			a.test.Logf("⌛ Waiting (%s) for flows: %s", wait, err)
-		}})
-	defer w.Cancel()
-
-retry:
-	since := a.flows[pod].lastTime()
-	if since.IsZero() {
-		since = a.started
-	}
-	flows, err := a.getFlows(ctx, hubbleClient, since, pod, podIP)
-	if err != nil || len(flows) == 0 {
-		if err == nil {
-			err = fmt.Errorf("no flows returned")
-		}
-		if err := w.Retry(err); err != nil {
-			if len(flows) == 0 {
-				retriesOnEmpty--
-				if retriesOnEmpty > 0 {
-					a.Infof("Retrying hubble flow retrieval on pod %q: %s", pod, err)
-					goto retryOnEmpty
-				}
-			}
-			a.Failf("Unable to retrieve flows of pod %q: %s", pod, err)
-			return err
-		}
-		goto retry
-	}
-
-	// append flows for the pod
-	flows = a.flows[pod].append(flows)
-	a.flows[pod] = flows
-
-	res := FlowRequirementResults{FirstMatch: -1, LastMatch: -1}
-	resFail := FlowRequirementResults{FirstMatch: -1, LastMatch: -1}
-	for i, req := range reqs {
-		offset := 0
-		var r FlowRequirementResults
-		for offset < len(flows) {
-			r = a.matchFlowRequirements(ctx, flows, offset, pod, &req)
-			if r.Failures > 0 {
-				resFail.Merge(&r)
+	// Only send readiness signal once.
+	var once sync.Once
+
+	for {
+		// Blocks, interruptable by context cancelation.
+		res, err := b.Recv()
+		if err != nil {
+			// Any of the following errors are expected and signal the end
+			// of the read loop.
+			if errors.Is(err, io.EOF) ||
+				errors.Is(err, context.Canceled) ||
+				errors.Is(err, context.DeadlineExceeded) {
+				return nil
 			}
-			// Check if successfully fully matched or no match for the first flow
-			if (!r.NeedMoreFlows && r.Failures == 0) || r.FirstMatch == -1 {
-				break
+
+			// Return gracefully on 'canceled' gRPC error.
+			if status.Code(err) == codes.Canceled {
+				return nil
 			}
-			// Try if some other flow instance would find both first and last required flows
-			offset = r.FirstMatch + 1
+
+			return fmt.Errorf("gRPC error: %w", err)
 		}
-		if r.NeedMoreFlows {
-			// Retry until timeout. On timeout, print the flows and
-			// consider it a failure
-			if err := w.Retry(errNeedMoreFlows); err == nil {
-				goto retry
+
+		switch r := res.GetResponseTypes().(type) {
+
+		case *observer.GetFlowsResponse_NodeStatus:
+			// Handle NodeStatus messages generated by Hubble peers, containing
+			// individual node readiness, unavailability, invalid filters etc.
+
+			switch r.NodeStatus.StateChange {
+			case relay.NodeState_NODE_CONNECTED:
+				// Received first connection event from a Hubble peer, tentatively
+				// notify the caller that traffic can be generated.
+				a.Debugf("Connected to Hubble node(s) %s", r.NodeStatus.NodeNames)
+				once.Do(func() { ready <- true })
```

Comment:
If I read this correctly, we return `ready` once at least one of the node is connected to? Shouldn't we return `ready` when all nodes are connected to?

Reply:
For posterity, how do we know (within the scope of handling a NODE_CONNECTED message) which nodes need to be connected to? This is mitigated by handling NODE_UNAVAILABLE messages. If any of those are received during the test, it will fail.

Reply:
With a `status` request you can get how many nodes are connected to and how many are unavailable.

Reply:
Ah, now I see `waitForRelay()` was added. Am I correct in assuming Hubble nodes replying to a status request is not the same as a `NODE_CONNECTED` reply in response to a flow follow operation? For example, Hubble nodes could come and go in between the status request and the follow operation, and a successful status request might not mean that all nodes will accept the follow filter.

To make this more robust, perhaps we should record the nodes returned from a status operation and expect to find the same nodes in the NODE_CONNECTED payload? Executing `waitForRelay()` might be a bit pointless otherwise.

Reply:
>  For example, Hubble nodes could come and go in between the status request and the follow operation, and a successful status response might not mean that all nodes will accept the follow filter.

correct.

> To make this more robust, perhaps we should record the nodes returned from a status operation and expect to find the same nodes in the NODE_CONNECTED payload?

also correct, checking `NODE_CONNECTED` payload would be more robust.